packages/neon-api/src/notifications/class.ts

```diff
@@ -0,0 +1,99 @@
+import { logging, settings } from "@cityofzion/neon-core";
+import WebSocket from "isomorphic-ws";
+import { NotificationMessage, CallbackFunction } from "./responses";
+import Subscription from "./subscription";
+const log = logging.default("api");
+
+interface ContractSubscriptions {
+  websocket: WebSocket;
+  callbacks: CallbackFunction[];
+}
+
+export class Notifications {
+  private url: string;
+
+  public get name(): string {
+    return `Notifications[${this.url}]`;
+  }
+
+  private subscriptions: Map<string | null, ContractSubscriptions>;
+
+  /**
+   * Create a new notification service that handles contract subscriptions
+   * @param url - URL of a notifications service.
+   */
+  public constructor(url: string) {
+    this.url = settings.networks[url]?.extra?.notifications ?? url;
+    this.subscriptions = new Map<string | null, ContractSubscriptions>();
+    log.info(`Created Notifications Provider: ${this.url}`);
+  }
+
+  /**
+   * Subscribe to event notifications of a specific contract
+   * @param contract - Hash of the contract (null for all contracts) to subscribe to
+   * @param callback - Function to call when a notification is received.
+   * @return Subscription object that can be used to cancel the subscription
+   */
+  public subscribe(
+    contract: string | null,
+    callback: CallbackFunction
+  ): Subscription {
+    if (this.subscriptions.has(contract)) {
+      this.subscriptions.get(contract)!.callbacks.push(callback);
+    } else {
+      const ws = new WebSocket(
+        this.url + (contract !== null ? "?contract=" + contract : "")
+      );
+      ws.onmessage = (event: WebSocket.MessageEvent) => {
+        for (const cb of this.subscriptions.get(contract)!.callbacks) {
+          cb(JSON.parse(event.data as string) as NotificationMessage);
+        }
+      };
+      this.subscriptions.set(contract, {
+        websocket: ws,
+        callbacks: [callback]
+      });
+    }
+    const unsubscribe = () => {
+      if (!this.subscriptions.has(contract)) {
+        // Needed because user might have called unsubscribeAll() before
+        return;
+      }
+      const subscriptions = this.subscriptions.get(contract)!;
+      const index = subscriptions.callbacks.indexOf(callback);
+      if (index === -1) {
+        // Could happen if unsubscribeAll() followed by subscribe() are called on the same contract
+        return;
+      }
+      if (subscriptions.callbacks.length > 1) {
+        subscriptions.callbacks.splice(index, 1);
+      } else {
+        this.unsubscribeContract(contract);
+      }
+    };
+    return new Subscription(contract, unsubscribe);
+  }
+
+  /**
+   * Unsubscribe all callbacks from a specific contract
+   * @param contract - Hash of the contract (null for all contracts) to unsubscribe callbacks from

```

Comment:
we will have to make it clear that the null string is a special subscription that listens to all contracts and `unsubscribeContract() !== unsubscribeAll()`.

Reply:
Change the comments in ca90ea9508e29003659ea4e1e1a3395b4aa90b07 to try to make that clearer, thoughts?